### ISSUE TITLE
Detail on softmax

### ISSUE THREAD

**Issue by DevinKreuzer** [graphdeeplearning/graphtransformer#4]
Great work!

I have a question concerning the implementation of softmax in the graph_transformer_edge_layer.py

When you define the softmax, you use the following function:

```
def exp(field):
    def func(edges):
        # clamp for softmax numerical stability
        return {field: torch.exp((edges.data[field].sum(-1, keepdim=True)).clamp(-5, 5))}
    return func
```

Shouldn't the attention weights/scores be scalars? From what I see, each head has an 8-dimensional score vector which you then compute .sum() on. The graph_transformer_layer.py layer does not have this .sum() function.
```
def scaled_exp(field, scale_constant):
    def func(edges):
        # clamp for softmax numerical stability
        return {field: torch.exp((edges.data[field] / scale_constant).clamp(-5, 5))}

    return func
```

Would appreciate any clarification on this :)

Best,
Devin

**Comment by vijaydwivedi75** [comment]
Hi @DevinKreuzer,

The `.sum()` is done here in graph_transformer_layer.py.
https://github.com/graphdeeplearning/graphtransformer/blob/3c83b4ba5e45a2e25bbefde1b35d88a27ca3cfb2/layers/graph_transformer_layer.py#L18-L19



> @DevinKreuzer: Shouldn't the attention weights/scores be scalars? From what I see, each head has an 8-dimensional score vector

- In graph_transformer_edge_layer.py, the process of injecting available edge features is chosen to be feature-dimension wise, i.e. implicit attention scores (per feature dimension) is multiplied with available edge features (per feature dimension), in Eqn. 12 of the paper, and implemented as:
https://github.com/graphdeeplearning/graphtransformer/blob/3c83b4ba5e45a2e25bbefde1b35d88a27ca3cfb2/layers/graph_transformer_edge_layer.py#L33-L34


- Eqn. 12 outputs a _d_-dim feature vector (say _d_ is the feature dimension). This _d_-dim edge feature vector is critical since its passed to the edge feature pipeline (to be maintained at every layer), starting from Eqn. 10, towards Eqns. 16-18 in the paper. In Eqn.11 the features of \hat{w}_{i, j} are summed across the _d_-dimensions to obtain scalars, which is the `.sum()` that you mention in your query.

https://github.com/graphdeeplearning/graphtransformer/blob/3c83b4ba5e45a2e25bbefde1b35d88a27ca3cfb2/layers/graph_transformer_edge_layer.py#L44-L48

Hope this helps for understanding the implementation.
Vijay

**Comment by vijaydwivedi75** [comment]
Closing the issue for now. Feel free to open for any (further) clarification.

**Comment by ZikangZhou** [comment]
Hi,

I want to confirm whether my understanding of equation 12 is correct. 

I understand equation 12 in this way: (Q h_i * K h_j / sqrt(d_k)) is a scalar, and (E e_ij) is a d_k-dim vector. Then a scalar multiplying a vector gives a d_k-dim vector.  In equation 11, this d_k-dim vector is transformed to a scalar by computing w_1+w_2+...+w_dk. Is it correct?
